openlibrary/plugins/openlibrary/js/page-barcodescanner/LazyBookCard.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Class for displaying a card for a book, and having it
+ * update as it makes requests for information about the book.
+ */
+import Promise from 'promise-polyfill'; // polyfill Promise support for IE11
+
+export default class LazyBookCard {
+    /**
+     * @param {LazyBookCardState} state
+     */
+    constructor(state) {
+        this.state = state;
+        /** @type {JQuery} */
+        this.ui = null;
+    }
+
+    /**
+     * Render the HTML
+     * @return {JQuery}
+     */
+    render() {
+        this.ui = $(`
+          <a class="lazy-book-card">
+            <div class="cover">
+              <img>
+            </div>
+            <div class="info">
+              <div class="title"></div>
+              <div class="byline"></div>
+              <div class="identifier"></div>
+            </div>
+          </a>
+        `);
+
+        // Reset state
+        const state = this.state;
+        this.state = {};
+        this.updateState(state);
+
+        return this.ui;
+    }
+
+    /**
+     * Update the state and the UI
+     * @param {LazyBookCardState} newState
+     */
+    updateState(newState) {
+        const oldState = this.state;
+        newState = Object.assign({}, oldState, newState);
+
+        if (this.ui) {
+            if (oldState.link != newState.link) {
+                this.ui.attr('href', newState.link);
+            }
+
+            if (oldState.coverSrc != newState.coverSrc) {
+                this.ui.find('.cover img').attr('src', newState.coverSrc);
+            }
+
+            const textFields = ['title', 'byline', 'identifier'];
+            for (let field of textFields) {
+                if (oldState[field] != newState[field]) {
+                    this.ui.find(`.${field}`).text(newState[field]);
+                }
+            }
+
+            const classFields = ['loading', 'errored'];
+            for (let field of classFields) {
+                if (oldState[field] != newState[field]) {
+                    this.ui.toggleClass(field, newState[field]);
+                }
+            }
+        }
+
+        this.state = newState;
+    }
+
+    /**
+     * @param {string} isbn
+     * @return {LazyBookCard}
+     */
+    static fromISBN(isbn) {
+        const cardEl = new LazyBookCard({
+            title: isbn,
+            loading: true,
+            link: `https://openlibrary.org/isbn/${isbn}`,
+        });
+
+        fetch(`https://openlibrary.org/isbn/${isbn}.json`).then(r => r.json())
```

Comment:
I wonder if we want to use the current server rather than openlibrary.org (in case dev instances have a discrepency; could this break tests e.g. on travis?)

Reply:
Fixed :+1: